staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -47,36 +50,65 @@ type openAPIAggregator struct {
 	// provided for dynamic OpenAPI spec
 	openAPIService *handler.OpenAPIService
 
-	// Aggregator's OpenAPI spec (holds apiregistration group).
-	aggregatorOpenAPISpec *spec.Swagger
+	contextMapper request.RequestContextMapper
 
-	// Local (in process) delegate's OpenAPI spec.
-	inProcessDelegatesOpenAPISpec *spec.Swagger
+	openAPIAggregationController *OpenAPIAggregationController
+}
 
-	contextMapper request.RequestContextMapper
+func (s *openAPIAggregator) addLocalSpec(spec *spec.Swagger, localHandler http.Handler, name, etag string) {
+	localApiService := apiregistration.APIService{}
+	localApiService.Name = name
+	s.openAPISpecs[name] = &openAPISpecInfo{
+		etag:       etag,
+		apiService: localApiService,
+		handler:    localHandler,
+		spec:       spec,
+	}
 }
 
-func buildAndRegisterOpenAPIAggregator(delegateHandler http.Handler, webServices []*restful.WebService, config *common.Config, pathHandler common.PathHandler, contextMapper request.RequestContextMapper) (s *openAPIAggregator, err error) {
+func buildAndRegisterOpenAPIAggregator(delegationTarget server.DelegationTarget, webServices []*restful.WebService,
+	config *common.Config, pathHandler common.PathHandler, contextMapper request.RequestContextMapper) (s *openAPIAggregator, err error) {
 	s = &openAPIAggregator{
 		openAPISpecs:  map[string]*openAPISpecInfo{},
 		contextMapper: contextMapper,
 	}
+	s.openAPIAggregationController = NewOpenAPIAggregationController(s)
 
-	// Get Local delegate's Spec
-	s.inProcessDelegatesOpenAPISpec, err = s.downloadOpenAPISpec(delegateHandler)
-	if err != nil {
-		return nil, err
+	i := 0
+	for delegate := delegationTarget; delegate != nil; delegate = delegate.NextDelegate() {
+		handler := delegate.UnprotectedHandler()
+		if handler == nil {
+			continue
+		}
+		delegateSpec, etag, _, err := s.downloadOpenAPISpec(handler, "")
+		if err != nil {
+			return nil, err
+		}
+		if delegateSpec == nil {
+			continue
+		}
+		// First spec is the base for all other specs. Keep all of its path
+		// e.g. /api/ from this spec. For any other spec after first one we
+		// should only be interested in things in /apis/ path.
+		if i != 0 {
```

Comment:
the /api assumption is not generally true for `i == 0`. Can we look for the first delegate which serves `/api` instead? All after that get filtered.